python/src/aiconfig/default_parsers/openai_vision.py

```diff
@@ -0,0 +1,427 @@
+                    system_prompt = system_prompt["content"]
+                resolved_system_prompt = resolve_system_prompt(
+                    prompt, system_prompt, params, aiconfig
+                )
+                completion_params["messages"].append(
+                    {"content": resolved_system_prompt, "role": "system"}
+                )
+
+            # Default to always use chat context
+            if not hasattr(prompt.metadata, "remember_chat_context") or (
+                hasattr(prompt.metadata, "remember_chat_context")
+                and prompt.metadata.remember_chat_context != False
+            ):
+                # handle chat history. check previous prompts for the same model. if same model, add prompt and its output to completion data if it has a completed output
+                for i, previous_prompt in enumerate(aiconfig.prompts):
+                    # include prompts upto the current one
+                    if previous_prompt.name == prompt.name:
+                        break
+
+                    if aiconfig.get_model_name(
+                        previous_prompt
+                    ) == aiconfig.get_model_name(prompt):
+                        # Add prompt and its output to completion data. Constructing this prompt will take into account available parameters.
+                        add_prompt_as_message(
+                            previous_prompt,
+                            aiconfig,
+                            completion_params["messages"],
+                            params,
+                        )
+        else:
+            # If messages are already specified in the model settings, then just resolve each message with the given parameters and append the latest message
+            for i in range(len(completion_params.get("messages"))):
+                prompt_content = completion_params["messages"][i]["content"]
+                if not prompt_content:
+                    continue
+                if isinstance(prompt_content, str):
+                    completion_params["messages"][i]["content"] = (
+                        resolve_prompt_string(
+                            prompt,
+                            params,
+                            aiconfig,
+                            prompt_content,
+                        )
+                    )
+                else:
+                    # If the prompt content is an array, then it's a ChatCompletionContentPartParam object
+                    for j in range(len(prompt_content)):
+                        if prompt_content[j].get("text"):
+                            prompt_content[j]["text"] = resolve_prompt_string(
+                                prompt,
+                                params,
+                                aiconfig,
+                                prompt_content[j]["text"],
+                            )
+
+        # Add in the latest prompt
+        add_prompt_as_message(
```

Comment:
Ok yea, this use case I see is going to make the input adapters pretty difficult when we start allowing previous chat history to included non-text inputs to remember. It'll have to be like a nested check for each prompt --> model --> model parser to allow for input and output verification and mapping to the message completion params

Reply:
We should probably pull chat serialization into the config api instead (but that's a larger refactor). So, in run_inference if remember_chat_history is true we do something like
```
history_completion_params = config.deserialize_chat_history(prompt)
```
And that implementation would iteratively call the parser .deserialize on each one.

And then in run_inference we instead have an array of completion params that work for the model, and the parser logic can piece together the messages from each?

i.e. instead of coupling parser with prompt management logic, we have the parser just handle an array of completion params and it can figure out how to construct the message from that array of params (would just be a map over the completion params and append all messages to an array)

Reply:
Hmmm, yea, I like having the abstraction between how each model parser deserializes it's own prompt, and then another model parser that connects them together if needed

Reply:
Instead of array of completion params, can be a conversation context, array of conversation messages (openai spec)